sylt-compiler/src/typechecker.rs

```diff
@@ -0,0 +1,717 @@
+        );
+        Ok(res)
+    }
+
+    fn uni_op(
+        &mut self,
+        span: Span,
+        val: &Expression,
+        op: fn(&Type) -> Type,
+        name: &str,
+    ) -> Result<Type, Vec<Error>> {
+        let val = self.expression(val)?;
+        let res = op(&val);
+        type_error_if_invalid!(
+            self,
+            res,
+            span,
+            TypeError::UniOp { val, op: name.into() }
+        );
+        Ok(res)
+    }
+
+    fn expression_union_or_errors<'a>(&mut self, expressions: impl Iterator<Item = &'a Expression>) -> Result<Type, Vec<Error>> {
+        let ty: Vec<Type> = expressions.map(|e| self.expression(e)).collect::<Result<Vec<Type>, Vec<Error>>>()?;
+        Ok(Type::maybe_union(ty.iter(), self.compiler.blobs.as_slice()))
+    }
+
+    fn expression(&mut self, expression: &Expression) -> Result<Type, Vec<Error>> {
+        use ExpressionKind as EK;
+        let span = expression.span;
+        let res = match &expression.kind {
+            EK::Add(a, b) => self.bin_op(span, a, b, op::add, "Addition")?,
+            EK::Sub(a, b) => self.bin_op(span, a, b, op::sub, "Subtraction")?,
+            EK::Mul(a, b) => self.bin_op(span, a, b, op::mul, "Multiplication")?,
+            EK::Div(a, b) => self.bin_op(span, a, b, op::div, "Division")?,
+            EK::AssertEq(a, b) | EK::Eq(a, b) | EK::Neq(a, b) => {
+                self.bin_op(span, a, b, op::eq, "Equality")?
+            }
+            EK::Gt(a, b) | EK::Gteq(a, b) | EK::Lt(a, b) | EK::Lteq(a, b) => {
+                self.bin_op(span, a, b, op::cmp, "Comparison")?
+            }
+
+            EK::Is(a, b) => self.bin_op(span, a, b, |a, b| Type::Ty, "Is")?,
+            EK::In(a, b) => {
+                let a = self.expression(a)?;
+                let b = self.expression(b)?;
+                // TODO(ed): Verify the order here
+                let ret = match (&a, &b) {
+                    (a, Type::List(b))
+                    | (a, Type::Set(b))
+                    | (a, Type::Dict(b, _)) => b.fits(a, self.compiler.blobs.as_slice()),
+                    _ => Err("".into()),
+                };
+                if let Err(msg) = ret {
+                    let err = Error::TypeError {
+                        kind: TypeError::BinOp { lhs: a, rhs: b, op: "Containment".into() },
+                        file: self.file(),
+                        span,
+                        message: if msg.is_empty() { None } else { Some(format!("because {}", msg)) },
+                    };
+                    return Err(vec![err]);
+                }
+                Type::Bool
+            }
+
+            EK::Neg(a) => self.uni_op(span, a, op::neg, "Negation")?,
+
+            EK::And(a, b) => self.bin_op(span, a, b, op::and, "Boolean and")?,
+            EK::Or(a, b) => self.bin_op(span, a, b, op::or, "Boolean or")?,
+            EK::Not(a) => self.uni_op(span, a, op::not, "Boolean not")?,
```

Comment:
I would call these logical and, or and not. But maybe they're only defined for booleans.

Reply:
They're only defined for booleans - and the function names match `common::Op::{And,Or,Not}` and `machine::op::{and,or,not}`. So I'd rather not change them...